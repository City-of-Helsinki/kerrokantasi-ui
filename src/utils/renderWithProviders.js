/* eslint react/prop-types: 0 */
import React from 'react';
import { Provider } from 'react-redux';
import { render } from '@testing-library/react';
import { LoginProvider } from 'hds-react';
import { IntlProvider } from 'react-intl';

import createStore from "../createStore";
import { userOidcConfig, apiTokenClientConfig } from './oidcConfig';

const loginProviderConfig = {
  userManagerSettings: userOidcConfig,
  apiTokensClientSettings: apiTokenClientConfig,
  sessionPollerSettings: { pollIntervalInMs: 10000 },
}

const renderWithProviders = (
  ui,
  { preloadedState = {}, locale = 'fi', history, store = createStore(preloadedState), ...renderOptions } = {},
) => {
  const Wrapper = ({ children }) => (
    <LoginProvider {...loginProviderConfig}>
      <IntlProvider locale={locale}>
        <Provider store={store}>{children}</Provider>
      </IntlProvider>
    </LoginProvider>
  );
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
};


export default renderWithProviders;
