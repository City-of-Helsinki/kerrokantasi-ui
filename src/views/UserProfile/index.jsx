/* eslint-disable react/forbid-prop-types */
import React from 'react';
import PropTypes from 'prop-types';
import { injectIntl, FormattedMessage } from 'react-intl';
import { connect } from 'react-redux';
import Helmet from 'react-helmet';
import { ControlLabel, FormControl, FormGroup } from 'react-bootstrap';

import { getUser } from '../../selectors/user';
import { fetchFavoriteHearings, fetchUserComments, removeHearingFromFavorites } from '../../actions/index';
import HearingCardList from '../../components/HearingCardList';
import UserComment from '../../components/UserComment/UserComment';
import getAttr from '../../utils/getAttr';
import getMessage from '../../utils/getMessage';
import Icon from '../../utils/Icon';
import LoadSpinner from '../../components/LoadSpinner';

// Default params when fetching favorite/followed hearings.
const PARAMS = { following: true };

class UserProfile extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      selectedHearing: '',
      commentCount: this.props.profile.comments.count || 0,
    };
  }

  componentDidMount() {
    const {
      userState: { userExists },
      user,
    } = this.props;
    if (userExists && user) {
      this.props.fetchComments();
      this.props.fetchFavorites(PARAMS);
    }
  }

  componentDidUpdate(prevProps) {
    if (prevProps.userState.userLoading && !this.props.userState.userLoading) {
      this.props.fetchComments();
      this.props.fetchFavorites(PARAMS);
    }
    if (!prevProps.user && this.props.user) {
      this.props.fetchComments();
      this.props.fetchFavorites(PARAMS);
    }
    if (prevProps.profile.comments.count !== this.props.profile.comments.count) {
      this.setCommentCount(this.props.profile.comments.count);
    }
    if (prevProps.user && prevProps.user.favorite_hearings !== this.props.user.favorite_hearings) {
      this.props.fetchFavorites(PARAMS);
    }
  }

  /**
   * Sets state.selectedHearing according to value and updates state.commentCount if needed.
   * @param {string} value - hearing id or empty string when 'all' hearings.
   */
  setSelectedHearing = (value) => {
    const {
      profile: { comments },
    } = this.props;
    let { count } = comments;
    if (value) {
      count = comments.uniqueHearings.find((hearing) => hearing.id === value).commentCount;
    }
    this.setState({ selectedHearing: value, commentCount: count });
  };

  /**
   * Updates state.commentCount with value.
   * @param {number} value
   */
  setCommentCount = (value) => {
    this.setState({ commentCount: value });
  };

  /**
   * Returns comments according to state.selectedHearing,
   * comments made in a specific hearing or all comments if !selectedHearing.
   * @returns {JSX.Element|*[]}
   */
  getUserComments() {
    const {
      profile: { comments },
      intl: { locale },
    } = this.props;
    const { selectedHearing } = this.state;

    return (
      <div className='row'>
        <div className='commentlist'>
          {comments.results.reduce((visibleComments, comment) => {
            if (!selectedHearing || selectedHearing === comment.hearing) {
              visibleComments.push(<UserComment comment={comment} key={comment.id} locale={locale} />);
            }
            return visibleComments;
          }, [])}
        </div>
      </div>
    );
  }

  /**
   * Calls HearingCardList which returns a HearingCard for each favorite hearing.
   * @returns {JSX.Element|*[]}
   */
  getHearingCards() {
    const {
      intl,
      profile: { favoriteHearings },
    } = this.props;

    return (
      <HearingCardList
        className='user-favorite'
        hearings={favoriteHearings.results}
        intl={intl}
        language={intl.locale}
        showCommentCount={false}
        unFavoriteAction={this.props.removeFromFavorites}
        userProfile
      />
    );
  }

  /**
   * Returns comment order selection component.
   * @returns {JSX.Element}
   */
  getCommentOrderSelect() {
    const ORDERING_CRITERIA = {
      CREATED_AT_DESC: '-created_at',
      CREATED_AT_ASC: 'created_at',
    };
    return (
      <div className='col-md-4'>
        <div className='order-wrapper'>
          <FormGroup controlId='sort-select'>
            <ControlLabel>
              <FormattedMessage id='sort'>{(txt) => txt}</FormattedMessage>
            </ControlLabel>
            <FormControl
              componentClass='select'
              onChange={(event) => this.props.fetchComments({ ordering: event.target.value })}
            >
              {Object.keys(ORDERING_CRITERIA).map((key) => (
                <FormattedMessage id={key} key={key}>
                  {(message) => <option value={ORDERING_CRITERIA[key]}>{message}</option>}
                </FormattedMessage>
              ))}
            </FormControl>
          </FormGroup>
        </div>
      </div>
    );
  }

  /**
   * Used to display an icon and text informing user that certain content was not found.
   * @param {string} messageID - passed to FormattedMessage
   * @returns {JSX.Element}
   */
  // eslint-disable-next-line class-methods-use-this
  getContentNotFound(messageID) {
    return (
      <div className='content-not-found'>
        <Icon name='search' size='2x' aria-hidden />
        <FormattedMessage id={messageID}>{(txt) => <p>{txt}</p>}</FormattedMessage>
      </div>
    );
  }

  /**
   * Returns the hearing selection component with options based on profile.comments.uniqueHearings array.
   * @returns {JSX.Element}
   */
  hearingCommentSelect() {
    const {
      profile: {
        comments: { uniqueHearings },
      },
      intl: { locale },
    } = this.props;
    const allText = getMessage('all', locale);

    return (
      <div className='col-md-8'>
        <div className='selection-wrappers'>
          <FormGroup controlId='hearing-select'>
            <ControlLabel>
              <FormattedMessage id='selectHearingComments'>{(txt) => txt}</FormattedMessage>
            </ControlLabel>

            <FormControl
              componentClass='select'
              onChange={(event) => {
                this.setSelectedHearing(event.target.value);
              }}
            >
              <option value=''>{allText}</option>
              {uniqueHearings.reduce((hearingOptions, hearing) => {
                let textValue = getAttr(hearing.data.title, locale);
                // Option text is limited to 100 chars because the width of the dropdown is based on the widest text.
                if (textValue.length > 100) {
                  // Check if text contains any dates like 20.11.2021
                  const containsDateAt = textValue.search(/(\d.*)/g);
                  // Check if text contains full stops/points in the last 15 chars before the 100th character
                  const pointsBetweenRange = [...textValue.matchAll(/[.]/g)].reduce((points, point) => {
                    if (point.index > 85 && point.index < 100) {
                      points.push(point.index);
                    }
                    return points;
                  }, []);
                  // If the text contains a point near the limit and doesnt contain any dates -> slice at point.
                  if (pointsBetweenRange.length > 0 && containsDateAt === -1) {
                    textValue = textValue.slice(0, pointsBetweenRange[0] + 1);
                  }
                  // If text length is still over 100 -> use first 96 characters of text + '...'
                  if (textValue.length > 100) {
                    textValue = `${textValue.slice(0, 96)}...`;
                  }
                }
                hearingOptions.push(
                  <option key={hearing.id} value={hearing.id}>
                    {textValue}
                  </option>,
                );
                return hearingOptions;
              }, [])}
            </FormControl>
          </FormGroup>
        </div>
      </div>
    );
  }

  render() {
    const {
      userState: { userLoading },
      user,
      intl,
      profile: { favoriteHearings, comments },
    } = this.props;

    if (userLoading || !user) {
      return <LoadSpinner />;
    }
    const { commentCount } = this.state;
    // True when favorites have been fetched and results array has content.
    const hearingsLoaded = Object.keys(favoriteHearings).length > 0 && favoriteHearings.results.length !== 0;
    // True when comments have been fetched and results array has content.
    const commentsLoaded = Object.keys(comments).includes('results') && comments.results.length !== 0;
    return (
      <div className='container user-profile'>
        <Helmet title={intl.formatMessage({ id: 'userInfo' })} />
        <div className='row'>
          <FormattedMessage id='userInfo'>{(txt) => <h1>{txt}</h1>}</FormattedMessage>
          <FormattedMessage id='favoriteHearings'>{(txt) => <h2>{txt}</h2>}</FormattedMessage>
        </div>
        <div className='row'>
          <div className='col-md-12'>
            {hearingsLoaded ? this.getHearingCards() : this.getContentNotFound('noFavoriteHearings')}
          </div>
        </div>
        <div className='row'>
          <FormattedMessage id='userAddedComments' values={{ n: commentCount }}>
            {(txt) => <h2>{txt}</h2>}
          </FormattedMessage>
          <div className='col-md-12'>
            {commentsLoaded ? (
              <div className='user-comments-wrapper'>
                <div className='row'>
                  {this.hearingCommentSelect()}
                  {this.getCommentOrderSelect()}
                </div>
                {this.getUserComments()}
              </div>
            ) : (
              this.getContentNotFound('noAddedComments')
            )}
          </div>
        </div>
      </div>
    );
  }
}

UserProfile.propTypes = {
  fetchComments: PropTypes.func,
  fetchFavorites: PropTypes.func,
  intl: PropTypes.object,
  profile: PropTypes.object,
  removeFromFavorites: PropTypes.func,
  user: PropTypes.object,
  userState: PropTypes.shape({
    userExists: PropTypes.bool,
    userLoading: PropTypes.bool,
  }),
};

const existsSelector = (state) => ({
  userExists: !state.oidc.isLoadingUser && state.oidc.user !== null,
  userLoading: state.oidc.isLoadingUser,
});

/**
 * Returns array with objects for each unique hearing that
 * the user has commented on.
 * @param state
 * @returns {*[]|*}
 */
const uniqueHearingsCommented = (state) => {
  if (!state.user.profile.comments) {
    return [];
  }
  const comments = state.user.profile.comments.results;
  if (comments) {
    // get unique hearing id's
    const uniqueIDs = comments.reduce((uniqueIDTotal, comment) => {
      if (!uniqueIDTotal.includes(comment.hearing)) {
        uniqueIDTotal.push(comment.hearing);
      }
      return uniqueIDTotal;
    }, []);
    // return array with objects for each unique hearing
    return uniqueIDs.reduce((uniqueHearings, uniqueID) => {
      const specificHearing = comments.find((element) => element.hearing === uniqueID);
      const hearingCommentCount = comments.reduce((totalCount, current) => {
        if (current.hearing === uniqueID) {
          return totalCount + 1;
        }
        return totalCount;
      }, 0);
      uniqueHearings.push({
        id: uniqueID,
        data: specificHearing.hearing_data,
        commentCount: hearingCommentCount,
      });
      return uniqueHearings;
    }, []);
  }
  return [];
};

/**
 * Returns object properties at state.user.profile[key]
 * @param {object} state
 * @param {string} key
 * @returns {*}
 */
const profileSelector = (state, key) => ({
  ...state.user.profile[key],
});

const profileCombiner = (state) => ({
  comments: {
    ...profileSelector(state, 'comments'),
    uniqueHearings: uniqueHearingsCommented(state),
  },
  favoriteHearings: profileSelector(state, 'favoriteHearings'),
});

const mapDispatchToProps = (dispatch) => ({
  fetchComments: (params) => dispatch(fetchUserComments(params)),
  fetchFavorites: (params) => dispatch(fetchFavoriteHearings(params)),
  removeFromFavorites: (slug, hearingId) => dispatch(removeHearingFromFavorites(slug, hearingId)),
});

const mapStateToProps = (state) => ({
  user: getUser(state),
  userState: existsSelector(state),
  profile: profileCombiner(state),
});

export { UserProfile as UnconnectedUserProfile };

export default connect(mapStateToProps, mapDispatchToProps)(injectIntl(UserProfile));
