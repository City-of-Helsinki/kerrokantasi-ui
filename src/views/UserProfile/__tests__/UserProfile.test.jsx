/* eslint-disable sonarjs/no-duplicate-string */
import React from 'react';
import { BrowserRouter } from 'react-router-dom';
import { screen, within } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import configureStore from 'redux-mock-store';
import { thunk } from 'redux-thunk';

import UserProfile from '..';
import { getIntlAsProp } from '../../../../test-utils';
import renderWithProviders from '../../../utils/renderWithProviders';

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

jest.mock('../../../api', () => ({
  get: jest.fn().mockResolvedValue({ ok: true, json: () => ({}) }),
  post: jest.fn().mockResolvedValue({ ok: true, json: () => ({}) }),
}));

const createUniqueHearing = (id, slug, title, count, closed, ...props) => ({
  id,
  commentCount: count,
  data: {
    slug,
    title: {
      fi: title,
    },
    closed,
  },
  created_at: '2015-11-16T09:25:37.625607Z',
  ...props,
});
const defaultUniqueHearings = [
  createUniqueHearing('hearingOne', 'slugOne', 'this str contains a full stop so it will be cut.'.repeat(3), 1, true),
  createUniqueHearing(
    'hearingTwo',
    'slugTwo',
    'FILLER LOREM IPSUM FILLER. this contains a date near the end so it is sliced 12.12.2020. LOREM IPSUM FILLER.',
    2,
    false,
  ),
  createUniqueHearing('hearingThree', 'slugThree', 'this is 88 characters.'.repeat(4), 3, false),
  createUniqueHearing(
    'hearingFour',
    'slugFour',
    'this is supposed to be a title that is over 100 characters.'.repeat(2),
    1,
    false,
  ),
];

const createComment = (id, slug, title, hearing, content, ...props) => ({
  id,
  hearing,
  content,
  n_votes: 0,
  created_by: null,
  created_at: '2021-11-16T09:25:37.625607Z',
  answers: [],
  hearing_data: {
    slug,
    title: {
      fi: title,
    },
    closed: false,
  },
  ...props,
});

const defaultComments = [
  createComment('1', 'slugOne', 'this str contains a full stop so it will be cut.'.repeat(3), 'first comment'),
  createComment(
    '2',
    'slugTwo',
    'FILLER LOREM IPSUM FILLER. this contains a date near the end so it is sliced 12.12.2020. LOREM IPSUM FILLER.',
    'second comment',
  ),
  createComment(
    '3',
    'slugTwo',
    'FILLER LOREM IPSUM FILLER. this contains a date near the end so it is sliced 12.12.2020. LOREM IPSUM FILLER.',
    'third comment',
  ),
  createComment('4', 'slugThree', 'this is 88 characters.'.repeat(4), 'fourth comment'),
  createComment('5', 'slugThree', 'this is 88 characters.'.repeat(4), 'fifth comment'),
  createComment('6', 'slugThree', 'this is 88 characters.'.repeat(4), 'sixth comment'),
  createComment(
    '7',
    'slugFour',
    'this is supposed to be a title that is over 100 characters.'.repeat(2),
    'seventh comment',
  ),
];

const defaultState = {
  user: {
    data: {
      favorite_hearings: ['hearingOne'],
    },
    profile: {
      comments: {
        count: defaultComments.length,
        results: defaultComments,
        uniqueHearings: defaultUniqueHearings,
      },
      favoriteHearings: {
        count: 2,
        results: [
          { id: 'hearingOne', slug: 'slugOne', title: 'this str contains a full stop so it will be cut.'.repeat(3) },
        ],
      },
    },
  },
  oidc: { isLoadingUser: false, user: {} },
  userState: {
    userExists: true,
    userLoading: false,
  },
};

const renderComponent = (storeOverride = false) => {
  const props = {
    intl: getIntlAsProp(),
    fetchComments: jest.fn(),
    fetchFavorites: jest.fn(),
    removeFromFavorites: jest.fn(),
  };

  const store = storeOverride || mockStore(defaultState);

  return renderWithProviders(
    <BrowserRouter>
      <UserProfile {...props} />
    </BrowserRouter>,
    { store },
  );
};

describe('<UserProfile />', () => {
  it('should render correctly', () => {
    renderComponent();
  });

  it('should fetchComments and fetchFavorites if userExists && user', () => {
    const store = mockStore(defaultState);

    renderComponent(store);

    const expected = [{ type: 'beginFetchUserComments' }, { type: 'beginFetchFavoriteHearings' }];

    expect(store.getActions()).toEqual(expected);
  });

  it('should not fetchComments and fetchFavorites if userExists && user is false', () => {
    const store = mockStore({ ...defaultState, user: { ...defaultState.user, data: null } });

    renderComponent(store);

    expect(store.getActions()).toEqual([]);
  });

  it('should not fetchComments & fetchFavorites if userLoading', async () => {
    const store = mockStore({
      ...defaultState,
      user: { ...defaultState.user, data: null },
      userState: { userExists: false, userLoading: true },
    });

    renderComponent(store);

    expect(await screen.findByTestId('load-spinner')).toBeInTheDocument();

    expect(store.getActions()).toEqual([]);
  });

  it('should have correct amount of options', async () => {
    renderComponent();

    const user = userEvent.setup();

    const controls = await screen.findByTestId('hearing-selects');
    const buttons = await within(controls).findAllByRole('button');
    const toggle = buttons[0];

    const select = toggle.nextSibling;

    await user.click(toggle);

    expect(select.children).toHaveLength(8);
  });

  it('should select hearing', async () => {
    renderComponent();

    const user = userEvent.setup();

    const controls = await screen.findByTestId('hearing-selects');
    const buttons = await within(controls).findAllByRole('button');
    const toggle = buttons[0];
    const select = toggle.nextSibling;

    await user.click(toggle);

    await user.click(select.children[2]);

    const commentList = await screen.findByTestId('commentlist');

    expect(commentList.children).toHaveLength(1);
  });

  it('should change order', async () => {
    renderComponent();

    const user = userEvent.setup();

    const controls = await screen.findByTestId('hearing-selects');
    const buttons = await within(controls).findAllByRole('button');
    const toggle = buttons[1];
    const select = toggle.nextSibling;

    const commentList = await screen.findByTestId('commentlist');

    expect(commentList.children).toHaveLength(7);

    await user.click(toggle);
    await user.click(select.children[1]);

    expect(commentList.children).toHaveLength(7);
  });

  it('should render no content found', async () => {
    const store = mockStore({
      ...defaultState,
      user: {
        ...defaultState.user,
        profile: {
          ...defaultState.user.profile,
          favoriteHearings: {
            count: 0,
            results: [],
          },
        },
      },
    });

    renderComponent(store);

    expect(screen.queryByText('noFavoriteHearings')).toBeInTheDocument();
  });

  it('should remove from favorites', async () => {
    const store = mockStore(defaultState);

    renderComponent(store);

    const user = userEvent.setup();

    const hearingCards = await screen.findByTestId('hearing-card-list');

    expect(hearingCards.children).toHaveLength(1);

    const removeFromFavorites = await within(hearingCards.children[0]).findByRole('button');

    await user.click(removeFromFavorites);

    const expected = [
      { type: 'beginFetchUserComments' },
      { type: 'beginFetchFavoriteHearings' },
      { type: 'receiveUserComments', payload: { data: {} } },
      { type: 'receiveFavoriteHearings', payload: { data: {} } },
      { type: 'beginRemoveHearingFromFavorites', payload: { hearingSlug: 'slugOne' } },
    ];

    expect(store.getActions()).toEqual(expected);
  });
});
