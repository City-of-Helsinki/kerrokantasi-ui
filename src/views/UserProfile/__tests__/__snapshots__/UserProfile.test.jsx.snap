// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UserProfile /> should render correctly 1`] = `
<div>
  <div
    class="container user-profile"
  >
    <div
      class="row"
    >
      <h1>
        userInfo
      </h1>
      <h2>
        favoriteHearings
      </h2>
    </div>
    <div
      class="row"
    >
      <div
        class="col-md-12"
      >
        <div
          class="hearing-card-list row"
          data-testid="hearing-card-list"
        >
          <div
            class="col-md-3 col-xs-12"
          >
            <div
              class="hearing-card user-favorite"
            >
              <div>
                <div
                  alt="firstFavoriteHearing"
                  class="mouse-only-click-element hearing-card-image"
                  style="background-image: url(test-file-stub);"
                />
              </div>
              <div
                class="hearing-card-content"
              >
                <h3
                  class="h4 hearing-card-title"
                >
                  <a
                    href="/undefined?headless=false"
                  >
                    firstFavoriteHearing
                  </a>
                </h3>
                <div
                  class="hearing-card-time expires"
                >
                  <span />
                </div>
                <div
                  class="hearing-card-labels clearfix"
                >
                  <div
                    aria-label="labelListTagsTitle"
                    class="hearing-list-item-labellist"
                  />
                </div>
                <div
                  class="favorite-button-wrapper"
                >
                  <button
                    aria-label="Poista suosikeista"
                    class="favorite-icon"
                    title="Poista suosikeista"
                    type="button"
                  >
                    <span
                      aria-hidden="true"
                      class="fa fa-heart"
                    />
                  </button>
                </div>
              </div>
            </div>
          </div>
          <div
            class="col-md-3 col-xs-12"
          >
            <div
              class="hearing-card user-favorite"
            >
              <div>
                <div
                  alt="secondFavoriteHearing"
                  class="mouse-only-click-element hearing-card-image"
                  style="background-image: url(test-file-stub);"
                />
              </div>
              <div
                class="hearing-card-content"
              >
                <h3
                  class="h4 hearing-card-title"
                >
                  <a
                    href="/undefined?headless=false"
                  >
                    secondFavoriteHearing
                  </a>
                </h3>
                <div
                  class="hearing-card-time expires"
                >
                  <span />
                </div>
                <div
                  class="hearing-card-labels clearfix"
                >
                  <div
                    aria-label="labelListTagsTitle"
                    class="hearing-list-item-labellist"
                  />
                </div>
                <div
                  class="favorite-button-wrapper"
                >
                  <button
                    aria-label="Poista suosikeista"
                    class="favorite-icon"
                    title="Poista suosikeista"
                    type="button"
                  >
                    <span
                      aria-hidden="true"
                      class="fa fa-heart"
                    />
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="row"
    >
      <h2>
        userAddedComments
      </h2>
      <div
        class="col-md-12"
      >
        <div
          class="user-comments-wrapper"
        >
          <div
            class="row"
          >
            <div
              class="col-md-8"
            >
              <div
                class="selection-wrappers"
              >
                <div
                  class="form-group"
                >
                  <label
                    class="control-label"
                    for="hearing-select"
                  >
                    selectHearingComments
                  </label>
                  <select
                    class="form-control"
                    data-testid="hearing-select"
                    id="hearing-select"
                  >
                    <option
                      value=""
                    >
                      Kaikki
                    </option>
                    <option
                      value="hearingOne"
                    >
                      this str contains a full stop so it will be cut.this str contains a full stop so it will be cut.
                    </option>
                    <option
                      value="hearingTwo"
                    >
                      FILLER LOREM IPSUM FILLER. this contains a date near the end so it is sliced 12.12.2020. LOREM I...
                    </option>
                    <option
                      value="hearingThree"
                    >
                      this is 88 characters.this is 88 characters.this is 88 characters.this is 88 characters.
                    </option>
                    <option
                      value="hearingFour"
                    >
                      this is supposed to be a title that is over 100 characters.this is supposed to be a title that i...
                    </option>
                  </select>
                </div>
              </div>
            </div>
            <div
              class="col-md-4"
            >
              <div
                class="order-wrapper"
              >
                <div
                  class="form-group"
                >
                  <label
                    class="control-label"
                    for="sort-select"
                  >
                    sort
                  </label>
                  <select
                    class="form-control"
                    id="sort-select"
                  >
                    <option
                      value="-created_at"
                    >
                      CREATED_AT_DESC
                    </option>
                    <option
                      value="created_at"
                    >
                      CREATED_AT_ASC
                    </option>
                  </select>
                </div>
              </div>
            </div>
          </div>
          <div
            class="row"
          >
            <div
              class="commentlist"
            >
              <div
                class="hearing-comment"
              >
                <div
                  class="hearing-comment__comment-wrapper"
                >
                  <div
                    class="hearing-comment-header"
                  >
                    <div
                      class="hearing-comment-publisher"
                    >
                      <span
                        class="hearing-comment-user"
                      >
                        <span
                          aria-hidden="true"
                          class="fa fa-user"
                        />
                      </span>
                      <span
                        class="hearing-comment-date"
                      >
                        3 vuotta sitten
                      </span>
                    </div>
                    <div
                      class="hearing-comment-status"
                    >
                      <div>
                        <span
                          class="label label-success"
                        >
                          <span
                            class="sr-only"
                          >
                            commentHearingStatus
                          </span>
                          openHearing
                        </span>
                      </div>
                      <a
                        href="/undefined?headless=false"
                      />
                    </div>
                  </div>
                  <div
                    class="hearing-comment-body"
                  >
                    <p>
                      first comment
                    </p>
                  </div>
                  <div
                    class="hearing-comment__images"
                  />
                </div>
              </div>
              <div
                class="hearing-comment"
              >
                <div
                  class="hearing-comment__comment-wrapper"
                >
                  <div
                    class="hearing-comment-header"
                  >
                    <div
                      class="hearing-comment-publisher"
                    >
                      <span
                        class="hearing-comment-user"
                      >
                        <span
                          aria-hidden="true"
                          class="fa fa-user"
                        />
                      </span>
                      <span
                        class="hearing-comment-date"
                      >
                        3 vuotta sitten
                      </span>
                    </div>
                    <div
                      class="hearing-comment-status"
                    >
                      <div>
                        <span
                          class="label label-success"
                        >
                          <span
                            class="sr-only"
                          >
                            commentHearingStatus
                          </span>
                          openHearing
                        </span>
                      </div>
                      <a
                        href="/undefined?headless=false"
                      />
                    </div>
                  </div>
                  <div
                    class="hearing-comment-body"
                  >
                    <p>
                      second comment
                    </p>
                  </div>
                  <div
                    class="hearing-comment__images"
                  />
                </div>
              </div>
              <div
                class="hearing-comment"
              >
                <div
                  class="hearing-comment__comment-wrapper"
                >
                  <div
                    class="hearing-comment-header"
                  >
                    <div
                      class="hearing-comment-publisher"
                    >
                      <span
                        class="hearing-comment-user"
                      >
                        <span
                          aria-hidden="true"
                          class="fa fa-user"
                        />
                      </span>
                      <span
                        class="hearing-comment-date"
                      >
                        3 vuotta sitten
                      </span>
                    </div>
                    <div
                      class="hearing-comment-status"
                    >
                      <div>
                        <span
                          class="label label-success"
                        >
                          <span
                            class="sr-only"
                          >
                            commentHearingStatus
                          </span>
                          openHearing
                        </span>
                      </div>
                      <a
                        href="/undefined?headless=false"
                      />
                    </div>
                  </div>
                  <div
                    class="hearing-comment-body"
                  >
                    <p>
                      third comment
                    </p>
                  </div>
                  <div
                    class="hearing-comment__images"
                  />
                </div>
              </div>
              <div
                class="hearing-comment"
              >
                <div
                  class="hearing-comment__comment-wrapper"
                >
                  <div
                    class="hearing-comment-header"
                  >
                    <div
                      class="hearing-comment-publisher"
                    >
                      <span
                        class="hearing-comment-user"
                      >
                        <span
                          aria-hidden="true"
                          class="fa fa-user"
                        />
                      </span>
                      <span
                        class="hearing-comment-date"
                      >
                        3 vuotta sitten
                      </span>
                    </div>
                    <div
                      class="hearing-comment-status"
                    >
                      <div>
                        <span
                          class="label label-success"
                        >
                          <span
                            class="sr-only"
                          >
                            commentHearingStatus
                          </span>
                          openHearing
                        </span>
                      </div>
                      <a
                        href="/undefined?headless=false"
                      />
                    </div>
                  </div>
                  <div
                    class="hearing-comment-body"
                  >
                    <p>
                      fourth comment
                    </p>
                  </div>
                  <div
                    class="hearing-comment__images"
                  />
                </div>
              </div>
              <div
                class="hearing-comment"
              >
                <div
                  class="hearing-comment__comment-wrapper"
                >
                  <div
                    class="hearing-comment-header"
                  >
                    <div
                      class="hearing-comment-publisher"
                    >
                      <span
                        class="hearing-comment-user"
                      >
                        <span
                          aria-hidden="true"
                          class="fa fa-user"
                        />
                      </span>
                      <span
                        class="hearing-comment-date"
                      >
                        3 vuotta sitten
                      </span>
                    </div>
                    <div
                      class="hearing-comment-status"
                    >
                      <div>
                        <span
                          class="label label-success"
                        >
                          <span
                            class="sr-only"
                          >
                            commentHearingStatus
                          </span>
                          openHearing
                        </span>
                      </div>
                      <a
                        href="/undefined?headless=false"
                      />
                    </div>
                  </div>
                  <div
                    class="hearing-comment-body"
                  >
                    <p>
                      fifth comment
                    </p>
                  </div>
                  <div
                    class="hearing-comment__images"
                  />
                </div>
              </div>
              <div
                class="hearing-comment"
              >
                <div
                  class="hearing-comment__comment-wrapper"
                >
                  <div
                    class="hearing-comment-header"
                  >
                    <div
                      class="hearing-comment-publisher"
                    >
                      <span
                        class="hearing-comment-user"
                      >
                        <span
                          aria-hidden="true"
                          class="fa fa-user"
                        />
                      </span>
                      <span
                        class="hearing-comment-date"
                      >
                        3 vuotta sitten
                      </span>
                    </div>
                    <div
                      class="hearing-comment-status"
                    >
                      <div>
                        <span
                          class="label label-success"
                        >
                          <span
                            class="sr-only"
                          >
                            commentHearingStatus
                          </span>
                          openHearing
                        </span>
                      </div>
                      <a
                        href="/undefined?headless=false"
                      />
                    </div>
                  </div>
                  <div
                    class="hearing-comment-body"
                  >
                    <p>
                      sixth comment
                    </p>
                  </div>
                  <div
                    class="hearing-comment__images"
                  />
                </div>
              </div>
              <div
                class="hearing-comment"
              >
                <div
                  class="hearing-comment__comment-wrapper"
                >
                  <div
                    class="hearing-comment-header"
                  >
                    <div
                      class="hearing-comment-publisher"
                    >
                      <span
                        class="hearing-comment-user"
                      >
                        <span
                          aria-hidden="true"
                          class="fa fa-user"
                        />
                      </span>
                      <span
                        class="hearing-comment-date"
                      >
                        3 vuotta sitten
                      </span>
                    </div>
                    <div
                      class="hearing-comment-status"
                    >
                      <div>
                        <span
                          class="label label-success"
                        >
                          <span
                            class="sr-only"
                          >
                            commentHearingStatus
                          </span>
                          openHearing
                        </span>
                      </div>
                      <a
                        href="/undefined?headless=false"
                      />
                    </div>
                  </div>
                  <div
                    class="hearing-comment-body"
                  >
                    <p>
                      seventh comment
                    </p>
                  </div>
                  <div
                    class="hearing-comment__images"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
